Inject the location into ValidationErrors

sirosen opened the pull request:
In order to ensure that errors are always unambiguous, add the `location` to the message data for ValidationErrors.
This is done by adding a new level of nesting to the error messages, putting errors under the location as a new top-level key.

Resolves #460

---

I'm putting this in a PR to evaluate it, but I'm not 100% comitted to this solution. I'm going to ask a couple of questions on #460 to try to better understand what the options are.

sirosen committed:
Inject the location into ValidationErrors

In order to ensure that errors are always unambiguous, add the
`location` to the message data for ValidationErrors.
This is done by adding a new level of nesting to the error messages,
putting errors under the location as a new top-level key.

Resolves #460

lafrech commented on tests/apps/flask_app.py:
```diff
@@ -212,4 +212,16 @@ def echo_use_kwargs_missing(username, **kwargs):
+    # on marshmallow2, validation errors for nested schemas can fail to encode:
+    # https://github.com/marshmallow-code/marshmallow/issues/493
+    # to workaround this, convert integer keys to strings
+    except TypeError:
```
I'd use `MARSHMALLOW_VERSION_INFO` here to

- ensure we are actually using MA2, not hiding another problem
- think of removing the code when dropping MA2

We can keep the try/except if you think it is better this way.

sirosen replied:
I'll change it right now. I think being able to flag it easily when dropping MA2 support is reason enough.

Also, you're right that catching a `TypeError` could hide other things (in theory, at least).

sirosen committed:
Dispatch on MA version in flask test app

Rather than catching an error which only occurs under marshmallow2,
explicitly test the version-info to see if it's running on ma2 or ma3.
This improves things in two ways:

1. makes it easier to find (and remove) this when MA2 support is
   dropped
2. ensures we aren't catching other TypeErrors when we handle this
   condition

sirosen commented:
Rebased so that this can merge cleanly. I think I said this clearly in #460, but to be clear: I no longer have any hesitation about merging this and putting it into 6.0.

lafrech commented:
Yes, I think this change makes total sense.

I guess it will mostly impact unit tests. I mean adapting to this change will probably be a few changes, if any, in application code, but mostly changes in unit tests asserting error messages. At least in my applications.

Let's go.

@sloria, I'd be happy to have your approval before releasing v6.
